Skip Hadoop classpath check if flink-shaded-hadoop jar exists in Flink lib directory

cutiechi opened the pull request:
<!--
Thanks for sending a pull request!

Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://kyuubi.readthedocs.io/en/master/contributing/code/index.html
  2. If the PR is related to an issue in https://github.com/apache/kyuubi/issues, add '[KYUUBI #XXXX]' in your PR title, e.g., '[KYUUBI #XXXX] Your PR title ...'.
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][KYUUBI #XXXX] Your PR title ...'.
-->

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug, and what versions are affected.
-->

This change addresses an issue where the Flink engine in Kyuubi would perform a Hadoop classpath check even when a ‎`flink-shaded-hadoop` jar is already present in the Flink ‎`lib` directory. In such cases, the check is unnecessary and may cause confusion or warnings in environments where the shaded jar is used instead of a full Hadoop classpath. By skipping the check when a ‎`flink-shaded-hadoop` jar exists, we improve compatibility and reduce unnecessary log output.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

The patch was tested by deploying Kyuubi with a Flink environment that includes a ‎`flink-shaded-hadoop` jar in the ‎`lib` directory and verifying that the classpath check is correctly skipped. Additional tests ensured that the check still occurs when neither the Hadoop classpath nor the shaded jar is present. Unit tests and manual verification steps were performed to confirm the fix.

### Was this patch authored or co-authored using generative AI tooling?
<!--
If a generative AI tooling has been used in the process of authoring this patch, please include
phrase 'Generated-by: ' followed by the name of the tool and its version.
If no, write 'No'.
Please refer to the [ASF Generative Tooling Guidance](https://www.apache.org/legal/generative-tooling.html) for details.
-->

No.

pan3793 commented on kyuubi-server/src/main/scala/org/apache/kyuubi/engine/flink/FlinkProcessBuilder.scala:
```diff
@@ -228,7 +228,21 @@ class FlinkProcessBuilder(
         extraCp.foreach(classpathEntries.add)
-        if (hadoopCp.isEmpty && extraCp.isEmpty) {
+
+        val flinkLibDir = Paths.get(flinkHome)
+          .resolve("lib")
+          .toFile
+
+        val hadoopJars = Option(flinkLibDir.listFiles(new FilenameFilter {
+          override def accept(dir: File, name: String): Boolean = {
+            name.toLowerCase.contains("hadoop") ||
+            name.toLowerCase.contains("flink-shaded-hadoop")
+          }
+        })).getOrElse(Array.empty)
+
+        val hasHadoopJar = hadoopJars.nonEmpty
```
simplify to
```suggestion
        val hasHadoopJar = Paths.get(flinkHome)
          .resolve("lib")
          .toFile
          .listFiles(new FilenameFilter {
            override def accept(dir: File, name: String): Boolean = {
              name.toLowerCase.startsWith("hadoop-client") ||
              name.toLowerCase.startsWith("flink-shaded-hadoop")
            }
          }).nonEmpty
```

cutiechi commented:
@pan3793 Please help me review again, thanks

pan3793 commented on kyuubi-server/src/main/scala/org/apache/kyuubi/engine/flink/FlinkProcessBuilder.scala:
```diff
@@ -228,9 +228,24 @@ class FlinkProcessBuilder(
+        }
+
+        if (!hasHadoopJar && hadoopCp.isEmpty && extraCp.isEmpty) {
+          warn(s"No Hadoop jar detected, and the conf of ${FLINK_HADOOP_CLASSPATH_KEY} " +
```
No hadoop client jars are found in ${flinkHome}/lib, and ...

update the following KyuubiException message too.

do we need to update docs too.

pan3793 commented on kyuubi-server/src/main/scala/org/apache/kyuubi/engine/flink/FlinkProcessBuilder.scala:
```diff
@@ -228,7 +228,21 @@ class FlinkProcessBuilder(
         hadoopCp.foreach(classpathEntries.add)
```
do you forget to delete this line?

cutiechi commented:
@pan3793 Thank you for your thorough review! I’ve updated the code and exception messages as suggested, and also added the relevant documentation. Please let me know if there are any further improvements needed.

cutiechi commented:
@pan3793 Hi, I noticed that this pull request has not been merged yet. Is there anything else I should modify or improve? Please let me know if any further changes are needed. Thank you!

cutiechi commented:
> @cutiechi there are test failures in `FlinkProcessBuilderSuite`, can you fix it?

Thanks for pointing it out! I will look into the test failures in `FlinkProcessBuilderSuite` and provide a fix as soon as possible.
